### DISCUSSION TITLE
[Question] Dynamic tag names

### DISCUSSION THREAD

**Discussion by matrei** [casid/jte#232] [Q&A]
Hi, I am looking for a way in JTE to pass in the tag name for a specific HTML element in a template.
In Thymeleaf there is a `th:tagName` attribute that will replace the defined tag name of an HTML element.

With `rootElementName` resolving to _span_:

```
<div th:tagName="${rootElementName}">Hello World!</div>
```

will render:

```
<span>Hello World!</span>
```

With JTE I tried:
```
@param String rootElementName
<${rootElementName}>Hello World!</${rootElementName}>
```
but this fails to compile with the error: _Expressions in HTML tag names are not allowed._

**Comment by kelunik** [comment]
I think this is better suited as an issue than discussions, but let's sort out some details first.

Could you describe the use cases where this would be useful?

**Comment by matrei** [comment]
Oh, sorry if I'm in the wrong forum.

Here are some use cases:

Configurable Tag Names:
Render different tags based on application configuration.
`<div th:tagName="${config.myTagName}"></div>`

Dynamic Tag Generation:
Generate tags dynamically based on data values (for example in a loop).
`<div th:each="item : ${items}" th:tagName="${item.tag}"></div>`

**Reply by edward3h** [reply]
What is the end goal of changing tag names? Could you achieve the effect you want a different way? For example, `<div>` with a `class=` set?

**Reply by matrei** [reply]
The end goal is to be have flexible control over the generated markup.

Sure, for some use cases, CSS could be a workaround. For example if it is sufficient to change the `display` property.

Coming from Thymeleaf, I could not find a way to accomplish the same thing in JTE, and hence my question.
I understand if it's not possible at the moment, I just wanted to check as I could not find a solution on my own.

**Reply by casid** [reply]
Just a thought about your original post:

```html
<div th:tagName="${condition ? 'a' : 'span'}"></div>
```

The idiomatic jte way would be instead:

```html
@if(condition)
  <a></a>
@else
  <span></span>
@endif
```

Now you get nice HTML completions from your IDE, for instance you probably need a `href` in the first case that you do not need (or would even be invalid HTML) in the second case.

**Reply by matrei** [reply]
Yes! That was a bad example that is better solved the way you show. :+1:
